### PR TITLE
Remove old Discovery Manager From Eclipse

### PR DESCRIPTION
**Pull request by stefaus**
This PR extends the `XMPPContactsService` by the retrieval of contact client features via the new `DiscoveryService` instead of the old `DiscoveryManager` and makes this information available as an attribute of an XMPPContact object.
This allows to remove the usage of the old `DiscoveryManager` outside of `saros.core`.

### CODE REVIEW

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/net/xmpp/contact/DiscoveryService.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/saros/saros/pullRequest?prid=4532156&bid=15255103)

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/net/xmpp/contact/DiscoveryService.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/saros/saros/pullRequest?prid=4532156&bid=15255103)

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/net/xmpp/contact/DiscoveryService.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/saros/saros/pullRequest?prid=4532156&bid=15255103)

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/net/xmpp/contact/DiscoveryService.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/saros/saros/pullRequest?prid=4532156&bid=15255103)

File: core/src/saros/net/xmpp/contact/DiscoveryService.java [hunk 1]
```diff
@@ -52,41 +54,52 @@ void connectionChanged(Connection connection) {
...
+    EnumSet<XMPPContact.Feature> features = EnumSet.noneOf(XMPPContact.Feature.class);
+
+    if (discoveryManager == null) return features;
     try {
       DiscoverInfo discoverInfo = discoveryManager.discoverInfo(fullJid);
-      return discoverInfo.containsFeature(feature);
+
+      for (Iterator<DiscoverInfo.Feature> it = discoverInfo.getFeatures(); it.hasNext(); ) {
+        String feature = it.next().getVar();
+        if (SarosConstants.XMPP_FEATURE_NAMESPACE.equals(feature)) {
```
Comment (srossbach) [inline core/src/saros/net/xmpp/contact/DiscoveryService.java]: The whole logic does not match the description of this class. You hardwired what features are queried. You have no option to query anything else and you are moving business logic even more into the network layer.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/net/xmpp/contact/DiscoveryService.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/saros/saros/pullRequest?prid=4532156&bid=15255103)

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/net/xmpp/contact/DiscoveryService.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/saros/saros/pullRequest?prid=4532156&bid=15255103)

File: core/src/saros/net/xmpp/contact/DiscoveryService.java (hunk 1 above)
Comment (stefaus) [reply core/src/saros/net/xmpp/contact/DiscoveryService.java]: For clarification I added a link to `XMPPContact.Feature` in the class description.

Yes I hardwired the xmpp names to resolve features here, and added an abstraction for features with `XMPPContact.Feature` enum. My idea is that later this enum will be part of a generic Contact Class and its the job of the contact service to resolve them.

How would you like to change this?

File: eclipse/src/saros/ui/views/SarosView.java [hunk 2]
```diff
@@ -453,17 +450,17 @@ public void menuAboutToShow(final IMenuManager manager) {
...
+                SelectionRetrieverFactory.getSelectionRetriever(XMPPContact.class).getSelection();
+            if (contacts.isEmpty()) return;
 
-            // FIXME dirty hack
-            final ContactPropertyTester tester = new ContactPropertyTester();
-            final boolean isOnline = tester.test(jid, "isOnline", null, null);
+            XMPPContact contact = contacts.get(0);
 
-            if (sarosSessionManager.getSession() == null && isOnline) {
+            // TODO OLD Behavior: here we check if contact is online (currently you can invite a
```
[the remark this answers is not in the document]
Comment (stefaus) [reply eclipse/src/saros/ui/views/SarosView.java]: The Wizard is checking the Saros Support, so which place do you mean?
Nevertheless, I added an additional remark.

File: eclipse/src/saros/ui/wizards/pages/ContactSelectionWizardPage.java [hunk 3]
```diff
@@ -64,24 +64,17 @@ public void contactSelectionChanged(ContactSelectionChangedEvent event) {
...
   /** This listener update the page completion if someone's presence changed. */
-  protected final DiscoveryManagerListener discoveryManagerListener =
-      new DiscoveryManagerListener() {
-        @Override
-        public void featureSupportUpdated(final JID jid, String feature, boolean isSupported) {
-
-          if (!SarosConstants.XMPP_FEATURE_NAMESPACE.equals(feature)) return;
+  private final IContactsUpdate contactsUpdate =
+      (contact, type) -> {
+        if (type != UpdateType.SAROS_SUPPORT) return;
```
[the remark this answers is not in the document]
Comment (stefaus) [reply eclipse/src/saros/ui/wizards/pages/ContactSelectionWizardPage.java]: It is not skipped. Every Feature Support change will trigger this UpdateType. A Server has also the Saros feature.

For clarification I renamed the enum to UpdateType.FEATURE_SUPPORT

File: core/src/saros/net/xmpp/contact/XMPPContact.java [hunk 4]
```diff
@@ -26,33 +27,35 @@
  */
 public class XMPPContact {
 
+  /** Possible Contact Features */
+  public enum Feature {
```
[the remark this answers is not in the document]
Comment (stefaus) [reply core/src/saros/net/xmpp/contact/XMPPContact.java]: My first thought was probably, I don't want something XMPP related in this enum and definitely no generic String comparison.
But yes as this is only a simple String identifier which is not really XMPP related I combined this to the enum `ResourceFeature`.

Hopefully this change also addresses @srossbach remark about hardwired features in DiscoveryService.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/net/ResourceFeature.java  3
- core/src/saros/net/xmpp/contact/DiscoveryService.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/saros/saros/pullRequest?prid=4532156&bid=15255103)

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/net/xmpp/contact/DiscoveryService.java  1
- core/src/saros/net/ResourceFeature.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/saros/saros/pullRequest?prid=4532156&bid=15255103)

**Comment by srossbach** [comment]
The whole logic would be better placed in the VersionManager as you have to request the version anyway. The whole DiscoveryStuff is a generic XMPP feature. However it is no longer used in a generic way (I guess the old DiscoveryManager should be deleted in the future) so there is no need for this.

This patch also pulls the "Server Part" deeper into the Core. Normally the IntelliJ and Eclipse Version has to be compiled against the Server (and even if it is only for a stupid string).

**Comment by stefaus** [comment]
So yes looking into the smack `ServiceDiscoveryManager` we could probably do this directly with the `VersionManager` as both need to send similar XMPP packets... 
But looking at the `VersionManager` I don't want to focus on refactoring this now... and first remove stuff from ui packages.

So I would stick with the `DiscoveryService` for now and refactor this stuff later.

> This patch also pulls the "Server Part" deeper into the Core.

I moved the value of `NAMESPACE_SERVER` from core:SarosConstants to this new enum in core.net. So could you explain where you see the problem here?

File: core/src/saros/net/ResourceFeature.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package saros.net;
+
+import java.util.Optional;
+
+/** Identifier of Saros Resource Features */
```
Comment (tobous) [inline core/src/saros/net/ResourceFeature.java]: Was the old reference to the XMPP specification (`<p>See also: XEP-0030: Service Discovery (http://xmpp.org/extensions/xep-0030.html)`) useful?

If so (and if this is the right position for it), you could consider adding it here as well as the constants still server the same purpose.

**Comment by saros-infrastructure** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/net/ResourceFeature.java  3
- core/src/saros/net/xmpp/contact/DiscoveryService.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/saros/saros/pullRequest?prid=4532156&bid=15255103)

File: core/src/saros/net/ResourceFeature.java (hunk 5 above)
Comment (stefaus) [reply core/src/saros/net/ResourceFeature.java]: This is a remark which is more related to the what the DiscoveryService is doing, but I don't think it's important to explicitly cite it.

**Pull request merged**